File: solutions/python/all-your-base/1/all_your_base.py
def decode(base: int, digits: list[int]) -> int:
    if base < 2:
        raise ValueError("input base must be >= 2")
    number = 0
    for digit in digits:
        if digit < 0 or base <= digit:
            raise ValueError("all digits must satisfy 0 <= d < input base")
        number = number * base + digit
    return number


def encode(base: int, number: int) -> list[int]:
    if base < 2:
        raise ValueError("output base must be >= 2")
    digits = []
    while number:
        digit = number % base
        number = number // base
        digits.append(digit)
    return list(reversed(digits)) if digits else [0]


def rebase(input_base: int, digits: list[int], output_base: int) -> list[int]:
    return encode(output_base, decode(input_base, digits))

File: solutions/go/all-your-base/1/all_your_base.go
package allyourbase

import (
	"errors"
	"slices"
)

var (
	errInputBase = errors.New("input base must be >= 2")
	errDigit     = errors.New(
		"all digits must satisfy 0 <= d < input base",
	)
	errOutputBase = errors.New("output base must be >= 2")
)

func decode(base int, digits []int) (int, error) {
	if base < 2 {
		return 0, errInputBase
	}
	number := 0
	for _, digit := range digits {
		if digit < 0 || base <= digit {
			return 0, errDigit
		}
		number = number*base + digit
	}
	return number, nil
}

func encode(base int, number int) ([]int, error) {
	if base < 2 {
		return nil, errOutputBase
	}
	digits := []int{}
	for number > 0 {
		digit := number % base
		number /= base
		digits = append(digits, digit)
	}
	if len(digits) == 0 {
		return []int{0}, nil
	}
	slices.Reverse(digits)
	return digits, nil
}

func ConvertToBase(
	inputBase int, inputDigits []int, outputBase int,
) ([]int, error) {
	number, err := decode(inputBase, inputDigits)
	if err != nil {
		return nil, err
	}
	return encode(outputBase, number)
}
